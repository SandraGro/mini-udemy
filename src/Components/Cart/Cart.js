import React, { useState, useEffect } from "react";
import { Container, Col, Row, Button, Modal } from "react-bootstrap";
import { Link, Route } from "react-router-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import * as icons from "@fortawesome/free-solid-svg-icons";
import "./Cart.scss";

function Cart(props) {
  const [totalOriginalPrice, setTotalOriginalPrice] = useState(null);
  const [totalDiscountedPrice, setTotalDiscountedPrice] = useState(null);

  const [isOpenBuyModal, setIsOpenBuyModal] = useState(false);
  console.log(props.user.cart);

  useEffect(() => {
    if (props.user.cart) {
      setTotalOriginalPrice(
        props.user.cart.reduce((total, course) => {
          return total + parseInt(course.originalPrice || 0);
        }, 0)
      );
      setTotalDiscountedPrice(
        props.user.cart.reduce((total, course) => {
          return total + parseInt(course.discountedPrice || 0);
        }, 0)
      );
    }
  }, [props.user.cart]);
  return (
    <>
      <Container>
        {props.user.cart ? (
          <p> {props.user.cart.length} courses in cart</p>
        ) : (
          ""
        )}
        <Row>
          <Col key={`courseItem-container`} sm={9}>
            <Row>
              {!props.user.cart || !props.user.cart.length ? (
                <p>Your cart is empty. Keep shopping to find a course!</p>
              ) : (
                props.user.cart.map((course, index) => (
                  <Col key={`courseItem-${index}`} sm={12}>
                    <Row>
                      <Col key={`courseItem-${index}`} sm={8}>
                        <div className="course-container">
                          <Link to={`/course/${course.slug}`}>
                            <div className="course-card-content">
                              <img
                                alt=""
                                src={course.thumbnail}
                                className="images thumbnail float-left"
                              />
                              <h6 className="card-title">{course.title}</h6>
                              <br />
                              <small>{course.author}</small>
                            </div>
                          </Link>
                        </div>
                      </Col>
                      <Col key={`courseLinks-${index}`} sm={2}>
                        <Button
                          className="link-button"
                          onClick={() =>
                            props.addCourseToCart(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            )
                          }
                        >
                          Remove
                        </Button>
                        <Button
                          className="link-button"
                          onClick={() => {
                            props.addCourseToSaveforLaterList(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                            props.addCourseToCart(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                          }}
                        >
                          Save for later
                        </Button>
                        <Button
                          className="link-button"
                          onClick={() => {
                            console.log(props.user.wishlistCourses, "user");
                            props.addCourseToWishlist(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses,
                              false
                            );
                            props.addCourseToCart(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                          }}
                        >
                          Move to wishlist
                        </Button>
                      </Col>
                      <Col sm={2}>
                        <Link
                          className="course-container"
                          to={`/course/${course.slug}`}
                        >
                          <div className="price">
                            <p className="discounted-price">
                              ${course.discountedPrice}{" "}
                              <span className="tag-icon">
                                <FontAwesomeIcon icon={icons.faTag} size="1x" />
                              </span>
                            </p>
                            <p className="original-price">
                              ${course.originalPrice}{" "}
                            </p>
                          </div>
                        </Link>
                      </Col>
                    </Row>
                    <hr />
                  </Col>
                ))
              )}
            </Row>
          </Col>
          <Col sm={3} className="align-top">
            <h6>Total:</h6>
            {props.user.cart ? (
              <h1>MXN${totalDiscountedPrice}</h1>
            ) : (
              <h1> There are no courses in the cart</h1>
            )}
            <Button
              variant="danger"
              size="lg"
              block
              onClick={() => setIsOpenBuyModal(true)}
            >
              Checkout
            </Button>
          </Col>
        </Row>
      </Container>
      <Modal show={isOpenBuyModal} onHide={() => setIsOpenBuyModal(false)}>
        <Modal.Header closeButton>
          <Modal.Title id="example-modal-sizes-title-lg">Summary</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <Container>
            <Row>
              <Col xs={8} md={8}>
                Original Price:
              </Col>
              <Col xs={4} md={4}>
                MXN${totalOriginalPrice}
              </Col>
            </Row>
            <Row>
              <Col xs={8} md={8}>
                Coupon discounts:
              </Col>
              <Col xs={4} md={4}>
                MMXN${totalDiscountedPrice - totalOriginalPrice}
              </Col>
            </Row>
            <hr></hr>
            <Row>
              <Col xs={8} md={8}>
                Total:
              </Col>
              <Col xs={4} md={4}>
                MMXN${totalDiscountedPrice}
              </Col>
            </Row>
          </Container>
        </Modal.Body>
        <Modal.Footer>
          <Button
            onClick={() => {
              setIsOpenBuyModal(false);
            }}
          >
            Complete Payment
          </Button>
        </Modal.Footer>
      </Modal>
      <Container>
        <p>Saved for later</p>
        <Row>
          <Col key={`courseItem-container`} sm={9}>
            <Row>
              {!(
                props.user.laterListCourses &&
                props.user.laterListCourses.length
              ) ? (
                <p>You haven't saved any courses for later.</p>
              ) : (
                props.user.laterListCourses.map((course, index) => (
                  <Col key={`courseItem-${index}`} sm={12}>
                    <Row>
                      <Col key={`courseItem-${index}`} sm={8}>
                        <div className="course-container">
                          <Link to={`/course/${course.slug}`}>
                            <div className="course-card-content">
                              <img
                                alt=""
                                src={course.thumbnail}
                                className="images thumbnail float-left"
                              />
                              <h6 className="card-title">{course.title}</h6>
                              <br />
                              <small>{course.author}</small>
                            </div>
                          </Link>
                        </div>
                      </Col>
                      <Col key={`courseLinks-${index}`} sm={2}>
                        <Button
                          className="link-button"
                          onClick={() =>
                            props.addCourseToSaveforLaterList(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            )
                          }
                        >
                          Remove
                        </Button>
                        <Button
                          className="link-button"
                          onClick={() => {
                            props.addCourseToSaveforLaterList(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                            props.addCourseToCart(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                          }}
                        >
                          Move to cart
                        </Button>
                      </Col>
                      <Col sm={2}>
                        <Link
                          className="course-container"
                          to={`/course/${course.slug}`}
                        >
                          <div className="price">
                            <p className="discounted-price">
                              ${course.discountedPrice}{" "}
                              <span className="tag-icon">
                                <FontAwesomeIcon icon={icons.faTag} size="1x" />
                              </span>
                            </p>
                            <p className="original-price">
                              ${course.originalPrice}{" "}
                            </p>
                          </div>
                        </Link>
                      </Col>
                    </Row>
                    <hr />
                  </Col>
                ))
              )}
            </Row>
          </Col>
        </Row>
      </Container>
      <Container>
        <p>Recently wishlisted</p>
        <Row>
          <Col key={`courseItem-container`} sm={9}>
            <Row>
              {!(
                props.user.wishlistCourses && props.user.wishlistCourses.length
              ) ? (
                <p>You haven't added any courses to your wishlist.</p>
              ) : (
                props.user.wishlistCourses.map((course, index) => (
                  <Col key={`courseItem-${index}`} sm={12}>
                    <Row>
                      <Col key={`courseItem-${index}`} sm={8}>
                        <div className="course-container">
                          <Link to={`/course/${course.slug}`}>
                            <div className="course-card-content">
                              <img
                                alt=""
                                src={course.thumbnail}
                                className="images thumbnail float-left"
                              />
                              <h6 className="card-title">{course.title}</h6>
                              <br />
                              <small>{course.author}</small>
                            </div>
                          </Link>
                        </div>
                      </Col>
                      <Col key={`courseLinks-${index}`} sm={2}>
                        <Button
                          className="link-button"
                          onClick={() =>
                            props.addCourseToWishlist(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            )
                          }
                        >
                          Remove
                        </Button>
                        <Button
                          className="link-button"
                          onClick={() => {
                            props.addCourseToWishlist(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                            props.addCourseToCart(
                              course.slug,
                              props.user,
                              props.setUser,
                              props.featuredCourses
                            );
                          }}
                        >
                          Move to cart
                        </Button>
                      </Col>
                      <Col sm={2}>
                        <Link
                          className="course-container"
                          to={`/course/${course.slug}`}
                        >
                          <div className="price">
                            <p className="discounted-price">
                              ${course.discountedPrice}{" "}
                              <span className="tag-icon">
                                <FontAwesomeIcon icon={icons.faTag} size="1x" />
                              </span>
                            </p>
                            <p className="original-price">
                              ${course.originalPrice}{" "}
                            </p>
                          </div>
                        </Link>
                      </Col>
                    </Row>
                    <hr />
                  </Col>
                ))
              )}
            </Row>
          </Col>
        </Row>
      </Container>
    </>
  );
}

export default Cart;
