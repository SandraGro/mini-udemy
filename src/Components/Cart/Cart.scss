.thumbnail {
    width: 150px;
}

button.link-button.btn.btn-primary  {
    color: #007791;
    background-color: transparent;
    font-weight: 400;
    border: none;
    display: block;
    text-decoration: none;
    font-size: 13px;
    padding: 0;
}

.course-card-content img{
    margin-right: 5px;
}

.price .discounted-price {
    margin-block-end: 0;
}

.original-price {
    text-decoration: line-through;
    color: #686f7a;
    font-size: 15px;
    font-weight: 200;
}

.discounted-price {
    color: #ec5252;
    font-size: 15px;
    font-weight: 700;
}