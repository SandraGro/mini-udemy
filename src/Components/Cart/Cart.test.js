import React from "react";
import { shallow, mount } from "enzyme";
import { BrowserRouter as Router } from "react-router-dom";
import toJson from "enzyme-to-json";
import courseList from "../../Fixtures/CourseList.json";
import Cart from "./Cart";

describe("Cart component tests", () => {
  const user = {
    cart: [],
    myCourses: [],
    id: "user01",
    username: "Sandra",
  };
  const setUser = function setUser(user) {
    return;
  };
  const featuredCourses = [];
  const addCourseToCart = function addCourseToCart(
    slug,
    user,
    setUser,
    featuredCourses
  ) {
    return;
  };
  const addCourseToSaveforLaterList = function addCourseToSaveforLaterList(
    slug,
    user,
    setUser,
    featuredCourses,
    deleteIfMatches = true
  ) {
    return;
  };
  const addCourseToWishlist = function addCourseToWishlist(
    slug,
    user,
    setUser,
    featuredCourses,
    deleteIfMatches = true
  ) {
    return;
  };

  it("should render the correct messages when the cart is empty", () => {
    const cart = shallow(
      <Cart
        user={user}
        addCourseToCart={addCourseToCart}
        addCourseToSaveforLaterList={addCourseToSaveforLaterList}
        addCourseToWishlist={addCourseToWishlist}
        setUser={setUser}
        featuredCourses={featuredCourses}
      />
    );
    expect(toJson(cart)).toMatchSnapshot();
  });

  it("should render the course details from the courses in cart", () => {
    const course = courseList["course01"];
    const testUser = {
      ...user,
      // const course = courseList["course01"];
      cart: [course],
    };
    const cart = mount(
      <Router>
        <Cart
          user={testUser}
          addCourseToCart={addCourseToCart}
          addCourseToSaveforLaterList={addCourseToSaveforLaterList}
          addCourseToWishlist={addCourseToWishlist}
          setUser={setUser}
          featuredCourses={featuredCourses}
        />
      </Router>
    );
    const courseContainer = cart.find("div.course-container");
    const pricesContainer = cart.find("a.course-container");
    expect(courseContainer.html()).toMatch(`${course.title}`);
    expect(courseContainer.html()).toMatch(`${course.author}`);
    expect(courseContainer.html()).toMatch(`${course.thumbnail}`);
    expect(pricesContainer.html()).toMatch(`${course.originalPrice}`);
    expect(pricesContainer.html()).toMatch(`${course.discountedPrice}`);
    expect(pricesContainer.html()).not.toMatch(`${course.category}`);
  });

  it("should render the total price correctly", () => {
    const courses = [
      courseList["course01"],
      courseList["course02"],
      courseList["course03"],
    ];
    let totalPrice = 0;
    for (let i in courses) {
      totalPrice += parseInt(courses[i].discountedPrice);
    }
    const testUser = {
      ...user,
      cart: courses,
    };
    const cart = mount(
      <Router>
        <Cart
          user={testUser}
          addCourseToCart={addCourseToCart}
          addCourseToSaveforLaterList={addCourseToSaveforLaterList}
          addCourseToWishlist={addCourseToWishlist}
          setUser={setUser}
          featuredCourses={featuredCourses}
        />
      </Router>
    );
    const pricesContainer = cart.find("div.total-price");
    expect(pricesContainer.html()).not.toMatch("MXN$0");
    expect(pricesContainer.html()).toMatch(`MXN$${totalPrice}`);
  });

  it("should render the checkout button when there is a course in the cart", () => {
    const courses = [
      courseList["course01"],
      courseList["course02"],
      courseList["course03"],
    ];

    const testUser = {
      ...user,
      cart: courses,
    };
    const cart = mount(
      <Router>
        <Cart
          user={testUser}
          addCourseToCart={addCourseToCart}
          addCourseToSaveforLaterList={addCourseToSaveforLaterList}
          addCourseToWishlist={addCourseToWishlist}
          setUser={setUser}
          featuredCourses={featuredCourses}
        />
      </Router>
    );
    const pricesContainer = cart.find("div.total-price");
    expect(pricesContainer.html()).toMatch("Checkout");
  });
});
