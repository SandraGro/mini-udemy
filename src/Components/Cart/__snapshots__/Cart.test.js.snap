// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart component tests should render the correct messages when the cart is empty 1`] = `
<Fragment>
  <Container
    fluid={false}
  >
    <p>
       
      0
       courses in cart
    </p>
    <Row
      noGutters={false}
    >
      <Col
        key="courseItem-container"
        sm={9}
      >
        <Row
          noGutters={false}
        >
          <p>
            Your cart is empty. Keep shopping to find a course!
          </p>
        </Row>
      </Col>
      <Col
        className="align-top total-price"
        sm={3}
      >
        <h6>
          Total:
        </h6>
        <h1>
          MXN$
        </h1>
        <Button
          active={false}
          block={true}
          disabled={false}
          onClick={[Function]}
          size="lg"
          variant="danger"
        >
          Checkout
        </Button>
      </Col>
    </Row>
  </Container>
  <Modal
    animation={true}
    autoFocus={true}
    backdrop={true}
    dialogAs={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "displayName": "ModalDialog",
        "render": [Function],
      }
    }
    enforceFocus={true}
    keyboard={true}
    onHide={[Function]}
    restoreFocus={true}
    show={false}
  >
    <ModalHeader
      closeButton={true}
      closeLabel="Close"
    >
      <ModalTitle
        id="example-modal-sizes-title-lg"
      >
        Summary
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <Container
        fluid={false}
      >
        <Row
          noGutters={false}
        >
          <Col
            md={8}
            xs={8}
          >
            Original Price:
          </Col>
          <Col
            md={4}
            xs={4}
          >
            MXN$
          </Col>
        </Row>
        <Row
          noGutters={false}
        >
          <Col
            md={8}
            xs={8}
          >
            Coupon discounts:
          </Col>
          <Col
            md={4}
            xs={4}
          >
            MMXN$
            0
          </Col>
        </Row>
        <hr />
        <Row
          noGutters={false}
        >
          <Col
            md={8}
            xs={8}
          >
            Total:
          </Col>
          <Col
            md={4}
            xs={4}
          >
            MMXN$
          </Col>
        </Row>
      </Container>
    </ModalBody>
    <ModalFooter>
      <Button
        active={false}
        disabled={false}
        onClick={[Function]}
        variant="primary"
      >
        Complete Payment
      </Button>
    </ModalFooter>
  </Modal>
  <Container
    fluid={false}
  >
    <p>
      Saved for later
    </p>
    <Row
      noGutters={false}
    >
      <Col
        key="courseItem-container"
        sm={9}
      >
        <Row
          noGutters={false}
        >
          <p>
            You haven't saved any courses for later.
          </p>
        </Row>
      </Col>
    </Row>
  </Container>
  <Container
    fluid={false}
  >
    <p>
      Recently wishlisted
    </p>
    <Row
      noGutters={false}
    >
      <Col
        key="courseItem-container"
        sm={9}
      >
        <Row
          noGutters={false}
        >
          <p>
            You haven't added any courses to your wishlist.
          </p>
        </Row>
      </Col>
    </Row>
  </Container>
</Fragment>
`;
